### DISCUSSION TITLE
Auxiliary/External Docs?

### DISCUSSION THREAD

**Discussion by Trequetrum** [sezanzeb/input-remapper#111] [Q&A] [answered]
First; this is an excellent little utility and it's the only way I can reasonably play Stardew Valley without a mouse. So thank you :)

I tried a lot of random stuff before I discovered that these events where what I was after:
```
e(EV_Rel, REL_WHEEL, 1)
e(EV_Rel, REL_WHEEL, -1)
```
This is basically the wheel macro without repeating. Even so, from the examples given there's not really a good way to know what can and cannot be done.

I wound up in `key-mapper/keymapper/utils.py`. I saw the imports for  `evdev.ecodes import EV_REL, REL_WHEEL, REL_HWHEEL` and thought "Huh, maybe they're relevant?"

I did end up at [python-evdev.readthedocs.io](https://python-evdev.readthedocs.io/en/latest/apidoc.html), but the API reference for ecodes doesn't say too much on the face of it.

Anyway, I'm sure there's somewhere you guys look for this sort of thing. It would be sweet if the docs included a link or two to external resources that users might find useful? Even the fact that `BTN_LEFT` is the mouse button and `Left` is the keyboard left button, for a newbie like me that makes it all confusing and a little bit of hand-holding wouldn't hurt! :) There's probably a wiki (or something) for linux event codes, but if you don't know, then you don't know it.

**Comment by sezanzeb** [comment, marked as answer]
I haven't come across a resource that explains all the possible names like REL_WHEEL and such.

You can use `sudo evtest` to find out type and code by using it on a device with a wheel. It would display something like
`Event: time 1621411092.319359, type 2 (EV_REL), code 8 (REL_WHEEL), value -1`. This is how I figured out all the relevant events.

There is this list: https://github.com/torvalds/linux/blob/master/include/uapi/linux/input-event-codes.h and this one: https://gist.github.com/sezanzeb/e29bae637b8a799ccf2490b8537487df

Names like "Left" are from xkb. Everything starting with KEY_, REL_, BTN_, ... are constants known to linux that are not affected by xkbs keyboard layout.
